lib/src/date-input/Calendar.tsx

```diff
@@ -0,0 +1,518 @@
+      } else {
+        setDateToFocus(selectedDate);
+      }
+    }
+  };
+
+  useEffect(() => {
+    document.getElementById(`day_${dateToFocus.get("date")}`)?.focus();
+  }, [dateToFocus]);
+
+  const handleDateSelect = (date: number, unit: UnitType) => {
+    setInnerDate((innerDate) => {
+      const newDate = innerDate.set(unit, date);
+      setSelectedDate(newDate);
+      onDateSelect(newDate);
+      return newDate;
+    });
+  };
+
+  const focusDate = (date: Dayjs) => {
+    if (innerDate.get("month") !== date.get("month") || innerDate.get("year") !== date.get("year")) {
+      setInnerDate((innerDate) => {
+        return innerDate.set("month", date.get("month")).set("year", date.get("year"));
+      });
+    }
+    setDateToFocus(date);
+  };
+
+  const handleDayKeyboardEvent = (event, day) => {
+    let dateToFocusTemp = innerDate.set("date", day);
+    switch (event.key) {
+      case "PageUp":
+        event.preventDefault();
+        event.shiftKey
+          ? (dateToFocusTemp = dateToFocusTemp.set("year", dateToFocusTemp.get("year") - 1))
+          : (dateToFocusTemp = dateToFocusTemp.set("month", dateToFocusTemp.get("month") - 1));
+        focusDate(dateToFocusTemp);
+        break;
+      case "PageDown":
+        event.preventDefault();
+        event.shiftKey
+          ? (dateToFocusTemp = dateToFocusTemp.set("year", dateToFocusTemp.get("year") + 1))
+          : (dateToFocusTemp = dateToFocusTemp.set("month", dateToFocusTemp.get("month") + 1));
+        focusDate(dateToFocusTemp);
+        break;
+      case "ArrowLeft":
+        event.preventDefault();
+        dateToFocusTemp = dateToFocusTemp.set("date", dateToFocusTemp.get("date") - 1);
+        focusDate(dateToFocusTemp);
+        break;
+      case "ArrowRight":
+        event.preventDefault();
+        dateToFocusTemp = dateToFocusTemp.set("date", dateToFocusTemp.get("date") + 1);
+        focusDate(dateToFocusTemp);
+        break;
+      case "ArrowUp":
+        event.preventDefault();
+        dateToFocusTemp = dateToFocusTemp.set("date", dateToFocusTemp.get("date") - 7);
+        focusDate(dateToFocusTemp);
+        break;
+      case "ArrowDown":
+        event.preventDefault();
+        dateToFocusTemp = dateToFocusTemp.set("date", dateToFocusTemp.get("date") + 7);
+        focusDate(dateToFocusTemp);
+        break;
+      case "Home":
+        event.preventDefault();
+        dateToFocus.get("day") !== 0
+          ? (dateToFocusTemp = dateToFocusTemp.day(1))
+          : (dateToFocusTemp = innerDate.date(day - 1).day(1));
+        focusDate(dateToFocusTemp);
+        break;
+      case "End":
+        event.preventDefault();
+        dateToFocusTemp.get("day") !== 0 && (dateToFocusTemp = dateToFocusTemp.day(7));
+        focusDate(dateToFocusTemp);
+        break;
+      case " ":
+        event.preventDefault();
+        handleDateSelect(day, "date");
+        break;
+    }
+  };
+
+  return (
+    <DatePicker ref={ref}>
```

Comment:
I think the JSX code of this component is huge. What about extracting some parts of it into separate components just to make the code cleaner? After reviewing the implementation I see three different modules based on `content`'s value:

1. Month Picker
2. Year Picker
3. Calendar
4. Date Picker

Date Picker would be the main parent component which would have all the logic, and the others would be representational components containing the different possible views of the picker and receive as props all the necessary values.

Reply:
When a component increases its side, it's recommended to separate it into smaller and more [pure components](https://beta.reactjs.org/learn/describing-the-ui#keeping-components-pure).